src/main/rx/nrf24_cx10.c

```diff
@@ -0,0 +1,373 @@
+    if (bindPacket) {
+        txId[0] = packet[1];
+        txId[1] = packet[2];
+        txId[2] = packet[3];
+        txId[3] = packet[4];
+        return true;
+    }
+    return false;
+}
+
+STATIC_UNIT_TESTED uint16_t convertToPwmUnsigned(const uint8_t* pVal)
+{
+    uint16_t ret = (*(pVal + 1)) & 0x7f; // mask out top bit which is used for a flag for the rudder
+    ret = (ret << 8) | *pVal;
+    return ret;
+}
+
+void cx10SetRcDataFromPayload(uint16_t *rcData, const uint8_t *payload)
+{
+    const uint8_t offset = (cx10Protocol == NRF24RX_CX10) ? 0 : 4;
+    rcData[NRF24_ROLL] = convertToPwmUnsigned(&payload[5 + offset]);  // aileron
+    rcData[NRF24_PITCH] = convertToPwmUnsigned(&payload[7 + offset]); // elevator
+    rcData[NRF24_THROTTLE] = convertToPwmUnsigned(&payload[9 + offset]);
+    rcData[NRF24_YAW] = convertToPwmUnsigned(&payload[11 + offset]);  // rudder
+}
+
+static void hopToNextChannel(void)
+{
+    ++rfChannelIndex;
+    if (rfChannelIndex >= RF_CHANNEL_COUNT) {
+        rfChannelIndex = 0;
+    }
+    NRF24L01_SetChannel(rfChannels[rfChannelIndex]);
+}
+
+// The hopping channels are determined by the txId
+STATIC_UNIT_TESTED void setHoppingChannels(const uint8_t* txId)
+{
+    rfChannelIndex = 0;
+    rfChannels[0] = 0x03 + (txId[0] & 0x0F);
+    rfChannels[1] = 0x16 + (txId[0] >> 4);
+    rfChannels[2] = 0x2D + (txId[1] & 0x0F);
+    rfChannels[3] = 0x40 + (txId[1] >> 4);
+}
+
+/*
+ * This is called periodically by the scheduler.
+ * Returns NRF24_RECEIVED_DATA if a data packet was received.
+ */
+nrf24_received_t cx10DataReceived(uint8_t *payload)
+{
+    nrf24_received_t ret = NRF24_RECEIVED_NONE;
+    switch (protocolState) {
+    case STATE_BIND:
+        if (NRF24L01_ReadPayloadIfAvailable(payload, payloadSize + 2)) {
+            XN297_UnscramblePayload(payload, payloadSize + 2);
+            const bool bindPacket = checkBindPacket(payload);
+            if (bindPacket) {
+                // set the hopping channels as determined by the txId received in the bind packet
+                setHoppingChannels(txId);
+                ret = NRF24_RECEIVED_BIND;
+                protocolState = STATE_ACK;
+                ackCount = 0;
+            }
+        }
+        break;
+    case STATE_ACK:
+        // transmit an ACK packet
+        ++ackCount;
+        // send out an ACK on the bind channel, required by deviationTx
+        payload[9] = 0x01;
+        NRF24L01_SetChannel(CX10_RF_BIND_CHANNEL);
+        NRF24L01_FlushTx();
+        XN297_WritePayload(payload, payloadSize);
+        NRF24L01_SetTxMode();// enter transmit mode to send the packet
+        delayMicroseconds(200);
+        // send out an ACK on each of the hopping channels, required by CX10 transmitter
+        for (uint8_t ii = 0; ii < RF_CHANNEL_COUNT; ++ii) {
+            NRF24L01_SetChannel(rfChannels[ii]);
+            while (NRF24L01_ReadReg(NRF24L01_17_FIFO_STATUS) & BV(NRF24L01_17_FIFO_STATUS_TX_FULL)) {
```

Comment:
Better wait for TX_EMPTY (to make sure that correct channel is used).

Datasheet is not helpful here. Channel is probably latched at start of TX mode.